### PR TITLE
Users have to explicitly specify group for in-development resource when kubectl 

### PR DESCRIPTION
**Pull request by janetkuo**
Partially fixes - #13929
Follow up #14461. 

`kubectl <operation> deployments` gives you error `experimental resource "deployments" should be specified explicitly; use experimental/deployments instead`

cc @lavalamp @caesarxuchao @bgrant0607

### CODE REVIEW

**Comment by JanetKuo** [comment]
I didn't add the check as advised [here](https://github.com/kubernetes/kubernetes/pull/14461/files#r41311591) because:
- if we return errors in `VersionAndKindForResource` when we see an implicit experimental resource, the `MultiRESTMapper` will try to find this resource in "/api/v1" and returns "resource not found"
- even if we change `MultiRESTMapper` to make it stop doing so, other commands like `kubectl create -f <file of experimental resource>` will fail and ask the users to specify experimental group.

**Comment by lavalamp** [comment]
The second commit looks good, but can we add a unit test or a line in test-cmd.sh? Thanks!

**Comment by caesarxuchao** [comment]
The second commit LGTM, too. Thanks.

**Comment by JanetKuo** [comment]
(#14461 is merged so you won't see its commit.)

I found some problems in the original code:
- Getting experimental resource with its shortcut (such as ds for daemonsets) will bypass the "implicit experimental resource" check in `ResourceTypeOrNameArgs` since resource builder doesn't recognize shortnames. i.e., users can do `kubectl get ds`. 
- It'll also make `kubectl get experimental/ds` not working.

So instead, I add the check in `ShortcutExpander`'s`VersionAndKindForResource`. Since it's in `ShortcutExpander` it won't affect `kubectl create -f <file of experimental resource>` like `DefaultRestMapper`. I also added tests in `test-cmd.sh`. 

PTAL. Thanks!

**Comment by caesarxuchao** [comment]
Thank you Janet, the changes SGTM.

Just a question,
> Since it's in ShortcutExpander it won't affect kubectl create -f <file of experimental resource> like DefaultRestMapper did
I'm confused why they were different. I tried on HEAD, DefaultRESTMapper.VersionAndKindForResouce() is not called when `kubectl create -f file`.
Thank you.

**Comment by caesarxuchao** [comment]
On second thought, I actually think it's better to do the check in `ResourceTypeOrNameArgs`, because we are sure that function is only called to when interpret command line args, while `VersionAndKindForResource` may have wider usage (though it's only reached by kubectl now). What do you think?

**Comment by JanetKuo** [comment]
@caesarxuchao thanks. You're right, DefaultRESTMapper.VersionAndKindForResouce() is not called when `kubectl create -f file`. I strikethrough those comments. 

As for where to add the check, per our discussion, we agreed that add it in `ShortcutExpander.VersionAndKindForResource` is the best choice since:
- we need to know the map from shortcut to resource name
- `ShortcutExpander` can only expand `ds` but not`experimental/ds` (so we should add the check at this level unless we change all experimental shortcut to `experimental/<original shortcut>`)
- we won't affect REST mapper's behavior

File: pkg/kubectl/kubectl.go [hunk 1]
```diff
@@ -77,8 +78,21 @@ type ShortcutExpander struct {
...
+	if len(parts) > 1 {
+		resource = fmt.Sprintf("%s/%s", parts[0], expandResourceShortcut(parts[1]))
+	}
+	// kind or kind-shortcut
+	if len(parts) == 1 {
+		resource = expandResourceShortcut(resource)
+	}
 	defaultVersion, kind, err = e.RESTMapper.VersionAndKindForResource(resource)
+	// kind that should be experimental/kind
+	if group, err := e.GroupForResource(resource); len(parts) == 1 && err == nil && group == exp {
```
Comment (caesarxuchao) [inline pkg/kubectl/kubectl.go]: How about checking len(parts)==1 before calling GroupForResource to save some function calls?

**Comment by caesarxuchao** [comment]
Thank you @JanetKuo, I agree ShortcutExpander maybe the best place to implement this logic.
LGTM except a nit.

File: pkg/kubectl/kubectl.go (hunk 1 above)
Comment (JanetKuo) [inline pkg/kubectl/kubectl.go]: Done

**Comment by caesarxuchao** [comment]
LGTM except a nit. Thanks.

File: hack/test-cmd.sh [hunk 2]
```diff
@@ -865,6 +866,33 @@ __EOF__
...
+  # Command
+  kubectl create -f examples/experimental/deployment.yaml
+  # Post-condition: deployment "nginx-deployment" is created
+  kube::test::get_object_assert experimental/deployment "{{range.items}}{{$id_field}}:{{end}}" 'nginx-deployment:'
+  # Clean up
+  kubectl delete experimental/deployment nginx-deployment
+  ### Get experimental resource via shortname
+  kubectl get experimental/hpa
+  ### Get experimental resource via shortname implicitly should fail
+  ERROR_FILE="/tmp/should-error"
```
Comment (caesarxuchao) [inline hack/test-cmd.sh]: Quinton pointed this out: https://github.com/kubernetes/kubernetes/pull/15325#discussion_r41575869

"If multiple of these run in parallel, they will trample on each other. You need to use a unique file name here."
"Actually ensure-temp-dir below might be what you're looking for."

Comment (JanetKuo) [inline hack/test-cmd.sh]: It's not a function. Will this run in parallel?

Comment (JanetKuo) [inline hack/test-cmd.sh]: Fixed.

File: hack/test-cmd.sh [hunk 3]
```diff
@@ -136,6 +136,7 @@ KUBE_API_VERSIONS="v1,experimental/v1alpha1" "${KUBE_OUTPUT_HOSTBIN}/kube-apiser
   --kubelet-port=${KUBELET_PORT} \
   --runtime-config=api/v1 \
   --cert-dir="${TMPDIR:-/tmp/}" \
+  --runtime_config="experimental/v1alpha1=true" \
```
Comment (JanetKuo) [inline hack/test-cmd.sh]: Is it okay to enable `experimental/v1alpha1` in `test-cmd.sh`?

File: hack/test-cmd.sh [hunk 4]
```diff
@@ -865,6 +866,33 @@ __EOF__
...
+  kube::test::get_object_assert experimental/deployment "{{range.items}}{{$id_field}}:{{end}}" 'nginx-deployment:'
+  # Clean up
+  kubectl delete experimental/deployment nginx-deployment
+  ### Get experimental resource via shortname
+  kubectl get experimental/hpa
+  ### Get experimental resource via shortname implicitly should fail
+  ERROR_FILE="${KUBE_TEMP}/should-error"
+  kubectl get hpa 1>&2 2> "${ERROR_FILE}" || true
+  if grep -q "error" "${ERROR_FILE}" && grep -q "should be specified explicitly" "${ERROR_FILE}"; then
+    echo "\"kubectl get hpa\" returns error as expected: $(cat ${ERROR_FILE})"
```
Comment (caesarxuchao) [inline hack/test-cmd.sh]: use kube::log::status instead?

**Comment by JanetKuo** [comment]
Changed "experimental" to "extensions"

**Comment by JanetKuo** [comment]
@k8s-bot test this

**Comment by JanetKuo** [comment]
@k8s-bot test this

**Comment by ncdc** [comment]
We'd prefer not to have to specify `extensions/<type>`, at least not for the resources in `extensions/v1beta1`, as they are resources that are being assigned a relatively stable level of compatibility.

@bgrant0607 @smarterclayton

**Comment by bgrant0607** [comment]
@ncdc we agree. The user should not have to specify `extensions/`. After 1.1, we need to rethink how to explicitly enable alpha-quality APIs in kubectl.

**Comment by lavalamp** [comment]
@JanetKuo thanks for doing this and I'm sorry that I didn't LGTM it before we changed our minds. ;)

**Pull request closed without merging**